### PR TITLE
[864] ADR 4 - non-integer trainee IDs

### PR DESCRIPTION
**Pull request by gpeng**
### Context

We have ADRs. We didn't have one for this decision.

### Changes proposed in this pull request

* Add ADR for non-integer trainee IDs.

### Guidance to review

Read the thing... 😁

### CODE REVIEW

**Comment by gpeng** [comment]
[Add ADR for URL secure token stuff](https://trello.com/c/NeBcfC6u/864-add-adr-for-url-secure-token-stuff)

**Review by defong — CHANGES_REQUESTED** [review]
Please add to register-trainee-teachers/docs/adr/index.md

File: docs/adr/0004-non-integer-trainee-ids.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# 4. Non-integer trainee IDs
+
+Date: 1 April 2021
```
Comment (defong) [inline docs/adr/0004-non-integer-trainee-ids.md:3]:
```suggestion
Date: 2021-04-01
```

https://github.com/DFE-Digital/register-trainee-teacher-data/blob/master/docs/adr/0001-record-architecture-decisions.md#L3

File: docs/adr/0004-non-integer-trainee-ids.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+Accepted
+
+## Context
+
+We have been using trainee IDs in the URL as per standard Rails convention/operation.
+
+It was felt that we should use non-integer IDs for a number of reasons:
+
+* remove predictability
+* interoperability with other systems without depending on DB ids
```
Comment (defong) [inline docs/adr/0004-non-integer-trainee-ids.md:16]:
```suggestion
* interoperability with other systems without depending on DB IDs
```

everywhere else is `IDs`

File: docs/adr/0004-non-integer-trainee-ids.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+#### Pros
+
+- in built uniqueness checks
+- framework support
+- not predictable
+- not as long as a full UUID
+
+#### Cons
+
+- record now has two 'ids'
```
Comment (defong) [inline docs/adr/0004-non-integer-trainee-ids.md:61]:
```suggestion
- record now has two 'IDs'
```

everywhere else is `IDs`

File: docs/adr/0004-non-integer-trainee-ids.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+## Decision
+
+We chose to use option 3 as it met the needs we had with the minimum of effort and avoided the really long URLs that 
+option 1 would have caused.
+
+## Consequences
+
+Trainee URLs are no longer predicatable and we have an effectively unique identifier.
```
Comment (defong) [inline docs/adr/0004-non-integer-trainee-ids.md:70]:
```suggestion
Trainee URLs are no longer predictable and we have an effective unique identifier.
```

File: docs/adr/template.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# NUMBER. TITLE
+
+Date: DATE
```
Comment (defong) [inline docs/adr/template.md:3]:
```suggestion
Date: YYYY-MM-DD
```

**Review by defong — CHANGES_REQUESTED** [review]

File: docs/adr/0004-non-integer-trainee-ids.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+It was felt that we should use non-integer IDs for a number of reasons:
+
+* remove predictability
+* interoperability with other systems without depending on DB IDs
+
+## Options
+
+### 1. UUIDs
+
+Use a 32 character UUID
```
Comment (defong) [inline docs/adr/0004-non-integer-trainee-ids.md:22]:
```suggestion
Use a 32 characters UUID
```

File: docs/adr/0004-non-integer-trainee-ids.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+- supported in the DB as a distinct type
+
+#### Cons
+
+- ugly URLs
+- long URLs
+- impossible to type
+- requires quite a lot of code changes
+
+### 2. Continue using integers IDs
```
Comment (defong) [inline docs/adr/0004-non-integer-trainee-ids.md:36]:
```suggestion
### 2. Continue using integer based IDs
```

**Review by defong — APPROVED** [review]

File: docs/adr/0004-non-integer-trainee-ids.md (hunk 6 above)
Comment (gpeng) [reply docs/adr/0004-non-integer-trainee-ids.md:22]: 'a 32 character' is correct...

Comment (defong) [reply docs/adr/0004-non-integer-trainee-ids.md:22]: ¯\_(ツ)_/¯ it's just poor grammar in any case

**Pull request merged**